File: leetcode/28.py
"""
Runtime: 218 ms, faster than 5.59% of Python3 online submissions for Implement strStr().
Memory Usage: 14.5 MB, less than 33.62% of Python3 online submissions for Implement strStr().
"""


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0
        if not haystack:
            return -1

        def checkIdx(i):
            return needle == haystack[i : i + len(needle)]

        for i in range(len(haystack)):
            if checkIdx(i):
                return i

        return -1

File: leetcode/28.cpp
// Runtime: 0 ms
// Memory Usage: 6.7 MB
class Solution {
 public:
  int strStr(string haystack, string needle) {
    for (string::const_iterator it = haystack.begin();
         it < haystack.end() - needle.size() + 1;
         ++it
    ) {
      string::const_iterator nit = needle.begin(), hit = it;
      while (nit != needle.end()
        && hit != haystack.end()
        && *hit == *nit
      ) {
        if (*nit != *hit) break;
        ++hit;
        ++nit;
      }
      if (nit == needle.end()) {
        return std::distance(
          static_cast<string::const_iterator>(haystack.begin()),
          it);
      }
    }
    return -1;
  }
};
